### PR TITLE
Add option to export non-public ceres headers.

### PR DESCRIPTION
**Pull request by schneith**

### CODE REVIEW

**Comment by schneith** [comment]
@simonlynen 
I would like to expose the internal functionality from ceres to build an external GN solver loop using the custom linear solver. Therefore I'm exporting the internal headers in the install step. Do you see any problems with this?

File: export_nonpublic.patch [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ 
++# Optionally install the non-public-headers.
++if (EXPORT_NONPUBLIC_HEADERS)
++  message("-- Export non-public headears.")
++  file(GLOB CERES_NON_PUBLIC_HDRS ${CMAKE_SOURCE_DIR}/internal/ceres/*.h)
++  install(FILES ${CERES_NON_PUBLIC_HDRS} DESTINATION include/ceres)
++else (EXPORT_NONPUBLIC_HEADERS)
++  message("-- Do not export non-public headers.")
++endif (EXPORT_NONPUBLIC_HEADERS)
++
```
Comment (tcies) [inline export_nonpublic.patch]: :fearful:

File: export_nonpublic.patch [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
++++ b/CMakeLists.txt
+@@ -717,6 +717,15 @@ install(FILES ${CERES_HDRS} DESTINATION include/ceres)
+ file(GLOB CERES_PUBLIC_INTERNAL_HDRS ${CMAKE_SOURCE_DIR}/include/ceres/internal/*.h)
+ install(FILES ${CERES_PUBLIC_INTERNAL_HDRS} DESTINATION include/ceres/internal)
+ 
++# Optionally install the non-public-headers.
++if (EXPORT_NONPUBLIC_HEADERS)
++  message("-- Export non-public headears.")
++  file(GLOB CERES_NON_PUBLIC_HDRS ${CMAKE_SOURCE_DIR}/internal/ceres/*.h)
++  install(FILES ${CERES_NON_PUBLIC_HDRS} DESTINATION include/ceres)
```
Comment (tcies) [inline export_nonpublic.patch]: Can you make this go in a more hidden place?

**Comment by tcies** [comment]
oqolgtm

File: export_nonpublic.patch (hunk 2 above)
Comment (schneith) [inline export_nonpublic.patch]: That might not work, as the (internal) headers rely on this path structure for their includes.

**Pull request merged**